### PR TITLE
Manage New Profiles

### PR DESCRIPTION
**Pull request by bonnichiwa**
## Description
Added new page to manage new profiles:
- `/profile/manage` => View and manage all profiles linked to a user
- `/profile/new` => Create a new profile

Updated existing routes (from PR1 https://github.com/hicommonwealth/commonwealth/pull/2418):

- `/profile/:address` to `/profile/:username` => View profile
- `/profile/:address/:edit` to `/profile/:username/edit` => Edit profile
- `/profile/a/:address/` => Redirects to `/profile/:username`

## How has this been tested?
TODO: Will test on staging

## Does this PR affect any server routes?
- [x] yes
- [ ] no

## If this PR affects server routes, what are the security implications?
- Updated routes to return Profile using username or address
- New route created to get all profiles and addresses belonging to a user
- New route created to delete profile (only when there are no addresses attached to it)
- New route created to create profile (only works if unique username is passed in)
- New route created to move address
- Modified route to delete address

## Have proper tags been added (for bug, enhancement, breaking change)?
- [x] yes

## Important Notes:
- A bunch of TODOs have been added to switch from using `/profile/a/:address` to `/profile/:username` once [PR4](https://www.notion.so/hicommonwealth/PR-4-Profiles-Conflict-Resolution-Jesse-to-Get-Numbers-on-Jake-Zak-b3b92e05b092470781569ce525dfffa0) has been merged. Currently there is no `username` on the `Profile` model, so a migration needs to happen to add a randomly generated `username` to every existing profile. New profiles will automatically have a randomly generated `username`. In the meantime, `profile/a/:address` just does a redirect to the linked profile (using the username) but most links will be broken as there is no username.

### CODE REVIEW

File: packages/commonwealth/server/models/profile.ts [hunk 1]
```diff
@@ -52,6 +53,7 @@ export default (
         primaryKey: true,
       },
       user_id: { type: dataTypes.INTEGER, allowNull: false },
+      username: { type: dataTypes.STRING, allowNull: false, unique: true },
```
Comment (jnaviask) [inline packages/commonwealth/server/models/profile.ts:56]: To enforce this, we will need to add generated profile names to each currently null "username" field on existing profiles. We will then need to define the "unique" constraint in the migration itself.

File: packages/commonwealth/server/routes/deleteProfile.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      Profiles: newProfiles,
+    },
+    {
+      where: {
+        id: req.user.id,
+      },
+    }
+  );
+
+  if (!updateProfileStatus && !updateUserStatus) {
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/deleteProfile.ts:60]: If I'm not mistaken, we should use a try-catch here (and in similar error-handling situations), as these calls will throw on error? + then we can halt control flow on the first failure. Not sure though, please confirm if possible.

File: packages/commonwealth/server/routes/moveAddress.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+  if (!req.body.newProfileId) {
+    return next(new Error(Errors.NoNewProfileProvided));
+  }
+
+  const { addressHash, oldProfileId, newProfileId } = req.body;
+
+  const address = await models.Address.findOne({
+    where: {
+      address: addressHash,
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/moveAddress.ts:34]: This is not a hash of the address, it is the address. We should rename to reflect that.

File: packages/commonwealth/server/routes/moveAddress.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    },
+  });
+
+  const oldProfile = await models.Profile.findOne({
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/moveAddress.ts:38]: We should validate at query time that each profile is owned by the current req.user, as a security consideration (i.e. not letting users move profiles around that aren't theirs).

File: packages/common-common/src/api/extApiTypes.ts [hunk 5]
```diff
@@ -160,4 +161,13 @@ export type PostRulesReq = {
 
 export type OnlyErrorResp = { error?: string | object };
 
+export type GetNewProfilesReq = {
+  jwt: string;
```
Comment (jnaviask) [inline packages/common-common/src/api/extApiTypes.ts:165]: Do we require the jwt = user login token for querying profiles? If it's exposed to the external API, we probably shouldn't, as that data is supposed to be used only on the commonwealth site.

**Review by jnaviask — COMMENTED** [review]
More comments on server-side code changes.

File: packages/commonwealth/server/models/profile.ts (hunk 1 above)
Comment (bonnichiwa) [reply packages/commonwealth/server/models/profile.ts:56]: Zak and I had discussed that adding generated usernames would be done in a later PR: https://www.notion.so/hicommonwealth/PR-4-Profiles-Conflict-Resolution-Jesse-to-Get-Numbers-on-Jake-Zak-b3b92e05b092470781569ce525dfffa0

@zakhap is this still the case?

File: packages/commonwealth/server/routes/deleteProfile.ts (hunk 2 above)
Comment (bonnichiwa) [reply packages/commonwealth/server/routes/deleteProfile.ts:60]: That's a good point, will add a try-catch to each of these calls.

File: packages/commonwealth/server/routes/moveAddress.ts (hunk 4 above)
Comment (bonnichiwa) [reply packages/commonwealth/server/routes/moveAddress.ts:38]: Nice! I will add that in

File: packages/common-common/src/api/extApiTypes.ts (hunk 5 above)
Comment (bonnichiwa) [reply packages/common-common/src/api/extApiTypes.ts:165]: Will remove

**Pull request merged**